system_tests/snap_sync_test.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2021-2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package arbtest
+
+import (
+	"context"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/offchainlabs/nitro/arbos/l2pricing"
+	"github.com/offchainlabs/nitro/util"
+	"math/big"
+	"os"
+	"testing"
+	"time"
+)
+
+func TestSnapSync(t *testing.T) {
+	ctx, cancelCtx := context.WithCancel(context.Background())
+	defer cancelCtx()
+
+	var transferGas = util.NormalizeL2GasForL1GasInitial(800_000, params.GWei) // include room for aggregator L1 costs
+
+	builder := NewNodeBuilder(ctx).DefaultConfig(t, true)
+	builder.L2Info = NewBlockChainTestInfo(
+		t,
+		types.NewArbitrumSigner(types.NewLondonSigner(builder.chainConfig.ChainID)), big.NewInt(l2pricing.InitialBaseFeeWei*2),
+		transferGas,
+	)
+	builder.Build(t)
+
+	// Added a delay, since BridgeBalance times out if the node is just created and not synced.
+	<-time.After(time.Second * 1)
+	builder.BridgeBalance(t, "Faucet", big.NewInt(1).Mul(big.NewInt(params.Ether), big.NewInt(10000)))
+
+	builder.L2Info.GenerateAccount("BackgroundUser")
+	// Sync node till batch count is 10
+	for {
+		tx := builder.L2Info.PrepareTx("Faucet", "BackgroundUser", builder.L2Info.TransferGas, big.NewInt(1), nil)
+		err := builder.L2.Client.SendTransaction(ctx, tx)
+		Require(t, err)
+		_, err = builder.L2.EnsureTxSucceeded(tx)
+		Require(t, err)
+		count, err := builder.L2.ConsensusNode.InboxTracker.GetBatchCount()
+		Require(t, err)
+		if count > 10 {
+			break
+		}
+	}
+	<-time.After(time.Second * 5)
+
+	batchCount, err := builder.L2.ConsensusNode.InboxTracker.GetBatchCount()
+	Require(t, err)
+	delayedCount, err := builder.L2.ConsensusNode.InboxTracker.GetDelayedCount()
+	Require(t, err)
+	// Last batch is batchCount - 1, so prev batch is batchCount - 2
+	prevBatchMetaData, err := builder.L2.ConsensusNode.InboxTracker.GetBatchMetadata(batchCount - 2)
+	Require(t, err)
+	prevMessage, err := builder.L2.ConsensusNode.TxStreamer.GetMessage(prevBatchMetaData.MessageCount - 1)
+	Require(t, err)
+	// Create a config with snap sync enabled and same database directory as the first node
+	nodeConfig := builder.nodeConfig
+	nodeConfig.SnapSync.Enabled = true
+	nodeConfig.SnapSync.BatchCount = batchCount
+	nodeConfig.SnapSync.DelayedCount = delayedCount
+	nodeConfig.SnapSync.PrevDelayedRead = prevMessage.DelayedMessagesRead
+	nodeConfig.SnapSync.PrevBatchMessageCount = uint64(prevBatchMetaData.MessageCount)
+	// Cleanup the message data, but keep the block state data.
+	// This is to simulate a snap sync environment where we’ve just gotten the block state but don’t have any messages.
+	err = os.RemoveAll(builder.l2StackConfig.ResolvePath("arbitrumdata"))
```

Comment:
here you have 1 node, you take it out, and bring it up again snap-sync style. The problem is we can't know that sync was correct because we have nothing to compare it to.
Instead, you should have 2 nodes:
1st sequencer that stays up all the test. You send all transactions to it.
2nd is just a node that syncs to the 1st - you wait until it's in sync
Then you take down/delete consensus dir/bring up using same execution dir - the 2nd node only
Then you add some new transactions to the sequencer, and test that the 2nd node can get synced with new blocks (by comparing these blocks from the 2 nodes)

Reply:
Done.